### ISSUE TITLE
`catch_unwind` panics due to failing debug assert when used during unwinding

### ISSUE THREAD

**Issue by tmiasko** [rust-lang/rust#68696]
In particular `catch_unwind` panics because of failing assert in https://github.com/rust-lang/rust/blob/266ecd6625060af304adc01f38773c2f013de1c5/src/libstd/panicking.rs#L289

For example: 

```rust

struct Catch;

impl Drop for Catch {
    fn drop(&mut self) {
        let _ = std::panic::catch_unwind(|| {});
    }
}

#[test]
#[should_panic]
fn it_works() {
    let catch = Catch;
    panic!("ups");
}
```

```shell
$ cargo test -Zbuild-std --target x86_64-unknown-linux-gnu
   Compiling panic v0.1.0 (/.../panic)
    Finished test [unoptimized + debuginfo] target(s) in 1.06s
     Running target/x86_64-unknown-linux-gnu/debug/deps/panic-118296df2d538b47

running 1 test
thread panicked while panicking. aborting.
error: test failed, to rerun pass '--lib'

Caused by:
  process didn't exit successfully: `/.../target/x86_64-unknown-linux-gnu/debug/deps/panic-118296df2d538b47` (signal: 4, SIGILL: illegal instruction)
```

**Comment by Mark-Simulacrum** [comment]
Cc @Amanieu 

I think that assert doesn't make much sense if I'm interpreting correctly, though I'm surprised it hasn't been hit before...

**Comment by tmiasko** [comment]
The current implementation seems to incorrectly assume that unwinding is not in progress when entering `catch_unwind`. Not caught before, since std is usually used without debug assertions (it definitely happens in actually existing code, what is unusual is that I am using `-Zbuild-std` and hence hitting the assertion).

There is a similar although a separate question if following should abort or not:

```rust
use std::panic::catch_unwind;

struct CatchPanic;

impl Drop for CatchPanic {
    fn drop(&mut self) {
        let _ = catch_unwind(|| {
            panic!("1");
        });
    }
}

fn main() {
    let _ = catch_unwind(|| {
        let _catch = CatchPanic;
        panic!("2");
    });
}
```

**Comment by Mark-Simulacrum** [comment]
Yes, the second code snippet should definitely abort - we abort on the second live panic, independent of whether it's going to be caught.
